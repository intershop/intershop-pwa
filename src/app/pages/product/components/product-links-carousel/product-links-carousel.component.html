<div *ngIf="links.productSKUs.length" class="product-list-container">
  <h2>{{ productLinkTitle }}</h2>

  <div class="product-list">
    <swiper [config]="swiperConfig">
      <div *ngFor="let sku of links.productSKUs" class="swiper-slide">
        <ish-product-item-container [productSku]="sku"></ish-product-item-container>
      </div>
    </swiper>
  </div>
</div>
