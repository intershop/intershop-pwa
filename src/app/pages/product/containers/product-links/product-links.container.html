<ng-container *ngIf="links$ | async as links">
  <ish-product-links-list
    [links]="links.upselling"
    [productLinkTitle]="'product.product_links.upselling.title' | translate"
  ></ish-product-links-list>
  <!---<ish-product-links-carousel [links]="links.crossselling"></ish-product-links-carousel>-->
</ng-container>
