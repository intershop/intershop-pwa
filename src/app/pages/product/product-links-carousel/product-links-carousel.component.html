<div *ngIf="links.productSKUs.length" class="product-list-container">
  <h2>{{ productLinkTitle }}</h2>

  <div class="product-list">
    <swiper [config]="swiperConfig">
      <div *ngFor="let sku of links.productSKUs" class="swiper-slide">
        <ish-product-item [productSku]="sku"></ish-product-item>
      </div>
    </swiper>
  </div>
</div>
