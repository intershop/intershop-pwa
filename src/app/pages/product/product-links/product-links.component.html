<ng-container *ngIf="links$ | async as links">
  <ng-container *ngIf="links.upselling?.productSKUs?.length">
    <ish-product-links-list
      [links]="links.upselling"
      [productLinkTitle]="'product.product_links.upselling.title' | translate"
    ></ish-product-links-list>
  </ng-container>
  <ng-container *ngIf="links.spareparts?.productSKUs?.length">
    <ish-product-links-list
      [links]="links.spareparts"
      [productLinkTitle]="'product.product_links.spareparts.title' | translate"
    ></ish-product-links-list>
  </ng-container>
  <ng-container *ngIf="links.crossselling?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.crossselling"
      [productLinkTitle]="'product.product_links.crossselling.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.accessory?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.accessory"
      [productLinkTitle]="'product.product_links.accessory.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.differentorderunit?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.differentorderunit"
      [productLinkTitle]="'product.product_links.differentorderunit.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.followup?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.followup"
      [productLinkTitle]="'product.product_links.followup.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.other?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.other"
      [productLinkTitle]="'product.product_links.other.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
  <ng-container *ngIf="links.replacement?.productSKUs?.length">
    <ish-product-links-carousel
      [links]="links.replacement"
      [productLinkTitle]="'product.product_links.replacement.title' | translate"
      [swiperConfig]="swiperConfig"
    ></ish-product-links-carousel>
  </ng-container>
</ng-container>
