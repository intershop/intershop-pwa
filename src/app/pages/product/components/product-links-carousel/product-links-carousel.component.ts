import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { SwiperConfigInterface } from 'ngx-swiper-wrapper';

import { ProductLinkView } from 'ish-core/models/product-links/product-links.model';

/**
 * The Product Link Carousel Component
 *
 * Displays the products which are assigned to a specific product link type as an carousel.
 * It uses the {@link ProductItemContainerComponent} for the rendering of products.
 *
 * @example
 * <ish-product-links-carousel [links]="links.crossselling" [productLinkTitle]="'product.product_links.crossselling.title' | translate"></ish-product-links-carousel>
 */
@Component({
  selector: 'ish-product-links-carousel',
  templateUrl: './product-links-carousel.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksCarouselComponent {
  /**
   * list of products which are assigned to the specific product link type
   */
  @Input() links: ProductLinkView;
  /**
   * title that should displayed for the specific product link type
   */
  @Input() productLinkTitle: string;
  /**
   * configuration for swiper carousel
   */
  @Input() swiperConfig: SwiperConfigInterface;
}
