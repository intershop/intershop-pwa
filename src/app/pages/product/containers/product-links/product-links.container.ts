import { ChangeDetectionStrategy, Component, Input, OnChanges } from '@angular/core';
import { Store, select } from '@ngrx/store';
import { Observable } from 'rxjs';

import { ProductLinksView } from 'ish-core/models/product-links/product-links.model';
import { LoadProductLinks, getProductLinks } from 'ish-core/store/shopping/products';

@Component({
  selector: 'ish-product-links-container',
  templateUrl: './product-links.container.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksContainerComponent implements OnChanges {
  @Input() sku: string;

  links$: Observable<ProductLinksView>;

  constructor(private store: Store<{}>) {}

  ngOnChanges() {
    if (this.sku) {
      this.store.dispatch(new LoadProductLinks({ sku: this.sku }));
      this.links$ = this.store.pipe(select(getProductLinks, { sku: this.sku }));
    }
  }
}
