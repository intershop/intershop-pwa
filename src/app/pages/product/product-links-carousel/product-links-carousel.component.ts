import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { SwiperOptions } from 'swiper';
import SwiperCore, { Navigation, Pagination } from 'swiper/core';

import { LARGE_BREAKPOINT_WIDTH, MEDIUM_BREAKPOINT_WIDTH } from 'ish-core/configurations/injection-keys';
import { ProductLinks } from 'ish-core/models/product-links/product-links.model';

SwiperCore.use([Navigation, Pagination]);

/**
 * The Product Link Carousel Component
 *
 * Displays the products which are assigned to a specific product link type as an carousel.
 * It uses the {@link ProductItemContainerComponent} for the rendering of products.
 *
 * @example
 * <ish-product-links-carousel [links]="links.crossselling" [productLinkTitle]="'product.product_links.crossselling.title' | translate"></ish-product-links-carousel>
 */
@Component({
  selector: 'ish-product-links-carousel',
  templateUrl: './product-links-carousel.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksCarouselComponent {
  /**
   * list of products which are assigned to the specific product link type
   */
  @Input() links: ProductLinks;
  /**
   * title that should displayed for the specific product link type
   */
  @Input() productLinkTitle: string;

  /**
   * configuration of swiper carousel
   * find possible parameters here: http://idangero.us/swiper/api/#parameters
   */
  swiperConfig: SwiperOptions;

  constructor(
    @Inject(LARGE_BREAKPOINT_WIDTH) largeBreakpointWidth: number,
    @Inject(MEDIUM_BREAKPOINT_WIDTH) mediumBreakpointWidth: number
  ) {
    this.swiperConfig = {
      direction: 'horizontal',
      keyboard: true,
      mousewheel: false,
      navigation: true,
      scrollbar: false,
      breakpoints: {
        0: {
          slidesPerView: 2,
          slidesPerGroup: 2,
        },
        [mediumBreakpointWidth]: {
          slidesPerView: 3,
          slidesPerGroup: 3,
        },
        [largeBreakpointWidth]: {
          slidesPerView: 4,
          slidesPerGroup: 4,
        },
      },
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
        clickableClass: 'swiper-pagination-clickable',
      },
    };
  }
}
