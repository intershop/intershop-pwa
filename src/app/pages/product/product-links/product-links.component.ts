import { ChangeDetectionStrategy, Component, Inject, Input, OnChanges } from '@angular/core';
import { SwiperConfigInterface } from 'ngx-swiper-wrapper';
import { Observable } from 'rxjs';

import { LARGE_BREAKPOINT_WIDTH } from 'ish-core/configurations/injection-keys';
import { ShoppingFacade } from 'ish-core/facades/shopping.facade';
import { ProductLinksView } from 'ish-core/models/product-links/product-links.model';

/**
 * The Product Link Container Component
 *
 * Collects all types of product links and prepares their rendering.
 * Product links can be displayed as product list (uses {@link ProductLinkListComponent}) or product carousel (uses  {@link ProductLinksCarouselComponent}).
 * For the carousel ngx-swiper-wrapper is used. Specific configuration for product link carousels have to be done here.
 *
 * @example
 * <ish-product-links-list [links]="links.upselling" [productLinkTitle]="'product.product_links.upselling.title' | translate"></ish-product-links-list>
 */
@Component({
  selector: 'ish-product-links',
  templateUrl: './product-links.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ProductLinksComponent implements OnChanges {
  /**
   * sku of the product whose product links should be rendered
   */
  @Input() sku: string;

  links$: Observable<ProductLinksView>;

  /**
   * configuration of swiper carousel
   * find possible parameters here: http://idangero.us/swiper/api/#parameters
   */
  swiperConfig: SwiperConfigInterface;

  constructor(private shoppingFacade: ShoppingFacade, @Inject(LARGE_BREAKPOINT_WIDTH) largeBreakpointWidth: number) {
    this.swiperConfig = {
      breakpoints: {
        [largeBreakpointWidth]: {
          slidesPerView: 2,
          slidesPerGroup: 2,
        },
      },
      pagination: {
        el: '.swiper-pagination',
        clickable: true,
        clickableClass: 'swiper-pagination-clickable',
      },
      slidesPerView: 4,
      slidesPerGroup: 4,
    };
  }

  ngOnChanges() {
    if (this.sku) {
      this.links$ = this.shoppingFacade.productLinks$(this.sku);
    }
  }
}
