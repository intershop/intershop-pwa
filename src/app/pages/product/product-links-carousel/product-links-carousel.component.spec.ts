import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { MockComponent } from 'ng-mocks';
import { SwiperModule } from 'ngx-swiper-wrapper';

import { ProductLinkView } from 'ish-core/models/product-links/product-links.model';
import { ProductItemComponent } from 'ish-shared/components/product/product-item/product-item.component';

import { ProductLinksCarouselComponent } from './product-links-carousel.component';

describe('Product Links Carousel Component', () => {
  let component: ProductLinksCarouselComponent;
  let fixture: ComponentFixture<ProductLinksCarouselComponent>;
  let element: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [SwiperModule],
      declarations: [MockComponent(ProductItemComponent), ProductLinksCarouselComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    const productLink = { productSKUs: ['sku'], categoryIds: ['catID'] } as ProductLinkView;

    fixture = TestBed.createComponent(ProductLinksCarouselComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    component.links = productLink;
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
