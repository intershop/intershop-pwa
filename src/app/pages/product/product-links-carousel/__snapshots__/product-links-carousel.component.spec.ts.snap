// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Links Carousel Component should be created 1`] = `
<div class="product-list-container">
  <h2></h2>
  <div class="product-list">
    <swiper
      ><div
        class="s-wrapper swiper swiper-container swiper-container-initialized swiper-container-horizontal"
        ng-reflect-disabled="false"
        ng-reflect-performance="false"
      >
        <div
          class="swiper-wrapper"
          style="webkit-transition-duration: 0ms; transition-duration: 0ms; webkit-transform: translate3d(NaNpx, 0px, 0px); transform: translate3d(NaNpx, 0px, 0px);"
        >
          <div class="swiper-slide">
            <ish-product-item ng-reflect-product-sku="sku"></ish-product-item>
          </div>
        </div>
        <div class="swiper-scrollbar" hidden=""></div>
        <div class="swiper-button-prev" hidden=""></div>
        <div class="swiper-button-next" hidden=""></div>
        <div class="swiper-pagination" hidden=""></div>
        <span class="swiper-notification" aria-live="assertive" aria-atomic="true"></span></div
    ></swiper>
  </div>
</div>
`;
